src/main/scala/lms/core/backend.scala

```diff
@@ -587,8 +587,10 @@ class GraphBuilderOpt extends GraphBuilder {
     case ("array_slice", List(as: Exp, Const(0), Const(-1))) => Some(as)
     case ("array_length", List(Def("NewArray", Const(n)::_))) =>
       Some(Const(n))
-    case ("array_length", List(Def("Array", List(Const(as: Array[_]))))) =>
+    case ("array_length", List(Def("Array", as))) =>
       Some(Const(as.length))
+    case ("array_length", List(Def("array_slice", List(x, s, e)))) =>
+      Some(reflect("-", e, s))
```

Comment:
fix array_length for array_slice (used to be hitting ???)

Reply:
The change of `array_length`'s first case seems changing the behavior. If the creation of `array_length` remains the same, now `as.length` is always 1, is that right？

Reply:
```  def Array[T:Manifest](xs: Rep[T]*): Rep[Array[T]] = {
    Wrap[Array[T]](Adapter.g.reflectMutable("Array", xs.map(Unwrap(_)):_*))
  }```

based on this code, we never store a list(const(array)), but a list of elements. 
the original code will never be triggered (unless it is array of array with just one element, and the result would be wrong 

Does this make sense?

Reply:
I see, it makes sense!